src/Orleans.Core/Async/BatchWorker.cs

```diff
@@ -157,88 +147,54 @@ private void CheckForMoreWork()
         /// <summary>
         /// Check if this worker is idle.
         /// </summary>
-        public bool IsIdle()
-        {
-            // no lock needed for reading volatile field
-            return currentWorkCycle == null;
-        }
+        public bool IsIdle() => currentWorkCycle == null;
 
         /// <summary>
         /// Wait for the current work cycle, and also the next work cycle if there is currently unserviced work.
         /// </summary>
-        /// <returns></returns>
-        public async Task WaitForCurrentWorkToBeServiced()
+        public Task WaitForCurrentWorkToBeServiced()
         {
-            Task<Task> waitfortasktask = null;
-            Task waitfortask = null;
-
             // Figure out exactly what we need to wait for
             lock (lockable)
             {
                 if (!moreWork)
                 {
                     // Just wait for current work cycle
-                    waitfortask = currentWorkCycle;
+                    return currentWorkCycle ?? Task.CompletedTask;
                 }
                 else
                 {
                     // we need to wait for the next work cycle
                     // but that task does not exist yet, so we use a promise that signals when the next work cycle is launched
-                    if (nextWorkCyclePromise == null)
-                    {
-                        nextWorkCyclePromise = new TaskCompletionSource<Task>(TaskCreationOptions.RunContinuationsAsynchronously);
-                    }
-
-                    waitfortasktask = nextWorkCyclePromise.Task;
+                    return nextWorkCycle ?? CreateNextWorkCyclePromise();
                 }
             }
+        }
 
-            // Do the actual waiting outside of the lock
-            if (waitfortasktask != null)
-            {
-                await await waitfortasktask;
-            }
-            else if (waitfortask != null)
-            {
-                await waitfortask;
-            }
+        private Task CreateNextWorkCyclePromise()
+        {
+            // it's OK to run any continuations synchrnously because this promise only gets signaled at the very end of CheckForMoreWork
+            nextWorkCyclePromise = new TaskCompletionSource<Task>();
```

Comment:
Prefer if we pass `TaskCreationOptions.RunContinuationsAsynchronously` for the reasons mentioned in the guidance here: https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#always-create-taskcompletionsourcet-with-taskcreationoptionsruncontinuationsasynchronously

Reply:
Looks good to me other than this, and possibly the same for the Task.Factory call (though with `TaskCreationOptions` instead). Hard to be sure that user code attaching continuations to these tasks will not end up causing a deadlock.

This structure of this class is not ideal in my eyes. I feel that a processor loop with a semaphore could be a cleaner approach. Still, this does look cleaner to me.

Reply:
I understand the guidance, but in this specific case it's safe and more optimal to potentially run continuations inline because it's signaled when CheckForMoreWork has finished and CheckForMoreWork is always scheduled asynchronously.

Reply:
`Task.Factory` does not have the same issues with continuations that TaskCompletionSource has because there is no explicit signaling of work completion.

Reply:
Sounds fine to me

Reply:
One last thing, though: was this PR in response to an issue you hit, or is it more speculative?

Reply:
No specific issue, I was just looking for `Task<Task>` uses and then noticed the problems in BatchWorker.Start method.